airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py

```diff
@@ -0,0 +1,184 @@
+        Example:
+            { "admin_graphql_api_id": "gid://shopify/Metafield/22533588451517" }
+            In this example the record could be identified by it's reference = ".../Metafield/..."
+        The property should be defined like:
+            record_identifier = "Metafield"
+        """
+
+    @property
+    @abstractmethod
+    def graph_edge(self) -> str:
+        """
+        Defines the final graph node name to fetch from.
+        """
+
+    @property
+    @abstractmethod
+    def graph_fields(self) -> Union[List[str], List[Any]]:
+        """
+        Defines the fields for final graph selection.
+        """
+
+    @abstractmethod
+    def resolve_query(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_query: str = None,
+        sort_key: str = None,
+    ) -> Operation:
+        """
+        Defines how query object should be resolved based on the base query selection.
+        """
+
+    def __new__(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_field: str = None,
+        start: str = None,
+        end: str = None,
+        sort_key: str = None,
+    ) -> str:
+        if not query_path:
+            raise NotImplementedError("The `query_path` is not defined.")
+        else:
+            # define query string
+            filter_query = f"{filter_field}:>='{start}' AND {filter_field}:<='{end}'" if filter_field else None
+            return self.resolve_query(self, query_path, filter_query, sort_key)
+
+
+class Metafields(ShopifyBulkQuery):
+
+    record_identifier = "Metafield"
+    graph_edge = "metafields"
+
+    # list of available fieds:
+    # https://shopify.dev/docs/api/admin-graphql/unstable/objects/Metafield
+    graph_fields = [
+        "id",
+        "namespace",
+        "value",
+        "key",
+        "description",
+        "createdAt",
+        "updatedAt",
+        "type",
+    ]
+
+    def resolve_query(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_query: str = None,
+        sort_key: str = None,
+    ) -> Operation:
+        """
+        Defines how query object should be constructed and resolved based on the base query selection.
+        """
+        if isinstance(query_path, list):
+            if len(query_path) == 2:
+                edges = self.get_edge(self, query_path[1], ["id", self.get_edge(self, self.graph_edge, self.graph_fields)])
+                query = self.get_query(self, query_path[0], edges, filter_query, sort_key)
+            if len(query_path) == 3:
+                edges = self.get_edge(
+                    self,
+                    query_path[1],
+                    ["id", self.get_edge(self, query_path[2], ["id", self.get_edge(self, self.graph_edge, self.graph_fields)])],
+                )
+                query = self.get_query(self, query_path[0], edges, filter_query, sort_key)
+            if len(query_path) == 1:
```

Comment:
there should be an else block that raises an exception if the length is unexpected